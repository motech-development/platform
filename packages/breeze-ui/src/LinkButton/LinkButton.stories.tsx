import { boolean, select, text, withKnobs } from '@storybook/addon-knobs';
import { MemoryRouter } from 'react-router-dom';
import BaseStyles from '../BaseStyles/BaseStyles';
import LinkButton from './LinkButton';

export default {
  component: LinkButton,
  decorators: [withKnobs],
};

const colour = {
  Danger: 'danger',
  Primary: 'primary',
  Secondary: 'secondary',
  Success: 'success',
};
const size = {
  Large: 'lg',
  Medium: 'md',
  Small: 'sm',
};

export const BasicLinkButton = {
  name: 'Basic link button',
  render: () => (
    <MemoryRouter>
      <BaseStyles />
      <LinkButton
        block={boolean('Block display', false)}
        colour={
          select('Colour', colour, 'primary') as
            | 'danger'
            | 'primary'
            | 'secondary'
            | 'success'
        }
        size={select('Size', size, 'md') as 'sm' | 'md' | 'lg'}
        to={text('To', '/home')}
      >
        {text('Title', 'Button')}
      </LinkButton>
    </MemoryRouter>
  ),
};
