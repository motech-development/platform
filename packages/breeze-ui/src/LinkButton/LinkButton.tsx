import React, { FC, memo } from 'react';
import { Link, LinkProps } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import BaseButton, {
  buttonTheme,
  IBaseButtonProps,
} from '../BaseButton/BaseButton';

const InternalLink: FC<ILinkButtonProps> = ({
  className,
  colour,
  block,
  ...rest
}) => (
  // eslint-disable-next-line react/jsx-props-no-spreading
  <Link className={className} {...rest} />
);

export interface ILinkButtonProps extends IBaseButtonProps, LinkProps {
  block?: boolean;
}

const LinkButton: FC<ILinkButtonProps> = ({
  block = false,
  colour = 'primary',
  size = 'md',
  ...rest
}) => (
  <ThemeProvider theme={buttonTheme}>
    <BaseButton
      as={InternalLink}
      block={block}
      colour={colour}
      size={size}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
    />
  </ThemeProvider>
);

export default memo(LinkButton);
