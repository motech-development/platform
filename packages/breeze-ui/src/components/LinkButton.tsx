import { ElementType } from 'react';
import { PolymorphicComponentProps } from 'react-polymorphic-box';
import TSize from '../utils/size';
import TTheme from '../utils/theme';
import BaseButton from './BaseButton';

interface ILinkButtonProps {
  block?: boolean;
  colour?: TTheme;
  size?: TSize;
}

export type TLinkButtonProps<E extends ElementType> = PolymorphicComponentProps<
  E,
  ILinkButtonProps
>;

const defaultElement = 'a';

const LinkButton = <C extends ElementType = typeof defaultElement>({
  block = false,
  colour = 'primary',
  size = 'md',
  ...rest
}: TLinkButtonProps<C>) => (
  <BaseButton
    as={defaultElement}
    block={block}
    colour={colour}
    size={size}
    // eslint-disable-next-line react/jsx-props-no-spreading
    {...rest}
  />
);

export default LinkButton;
