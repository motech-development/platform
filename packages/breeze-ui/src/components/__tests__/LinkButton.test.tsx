import { render } from '@testing-library/react';
import LinkButton from '../LinkButton';

describe('LinkButton', () => {
  describe('size', () => {
    it('should render when size is sm', () => {
      const { container } = render(
        <LinkButton href="/somewhere" size="sm">
          Hello
        </LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render when size is md', () => {
      const { container } = render(
        <LinkButton href="/somewhere">Hello</LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render when size is lg', () => {
      const { container } = render(
        <LinkButton href="/somewhere" size="lg">
          Hello
        </LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('theme', () => {
    it('should have the correct danger styles', () => {
      const { container } = render(
        <LinkButton href="/somewhere" colour="danger">
          Hello
        </LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct primary styles', () => {
      const { container } = render(
        <LinkButton href="/somewhere" colour="primary">
          Hello
        </LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct secondary styles', () => {
      const { container } = render(
        <LinkButton href="/somewhere" colour="secondary">
          Hello
        </LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct success styles', () => {
      const { container } = render(
        <LinkButton href="/somewhere" colour="success">
          Hello
        </LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct warning styles', () => {
      const { container } = render(
        <LinkButton href="/somewhere" colour="warning">
          Hello
        </LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('block', () => {
    it('should not be displayed as a block element by default', () => {
      const { container } = render(
        <LinkButton href="/somewhere">Hello</LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should not be displayed as a block element', () => {
      const { container } = render(
        <LinkButton href="/somewhere" block={false}>
          Hello
        </LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should be displayed as an block element', () => {
      const { container } = render(
        <LinkButton block href="/somewhere">
          Hello
        </LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct padding when sm', () => {
      const { container } = render(
        <LinkButton block href="/somewhere" size="sm">
          Hello
        </LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct padding when md', () => {
      const { container } = render(
        <LinkButton block href="/somewhere" size="md">
          Hello
        </LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct padding when lg', () => {
      const { container } = render(
        <LinkButton block href="/somewhere" size="lg">
          Hello
        </LinkButton>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
