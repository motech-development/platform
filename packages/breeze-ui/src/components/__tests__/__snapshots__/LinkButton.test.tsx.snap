// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkButton block should be displayed as an block element 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-2 px-4 text-base w-full hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;

exports[`LinkButton block should have the correct padding when lg 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-4 px-6 text-lg w-full hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;

exports[`LinkButton block should have the correct padding when md 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-2 px-4 text-base w-full hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;

exports[`LinkButton block should have the correct padding when sm 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-0 px-2 text-sm w-full hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;

exports[`LinkButton block should not be displayed as a block element 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-2 px-4 text-base hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;

exports[`LinkButton block should not be displayed as a block element by default 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-2 px-4 text-base hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;

exports[`LinkButton size should render when size is lg 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-4 px-6 text-lg hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;

exports[`LinkButton size should render when size is md 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-2 px-4 text-base hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;

exports[`LinkButton size should render when size is sm 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-0 px-2 text-sm hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;

exports[`LinkButton theme should have the correct danger styles 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-red-600 text-white py-2 px-4 text-base hover:bg-red-700 focus:ring-red-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;

exports[`LinkButton theme should have the correct primary styles 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-2 px-4 text-base hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;

exports[`LinkButton theme should have the correct secondary styles 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-gray-100 hover:bg-gray-300 focus:ring-gray-500 text-gray-600 py-2 px-4 text-base hover:bg-gray-700 focus:ring-gray-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;

exports[`LinkButton theme should have the correct success styles 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-green-600 text-white py-2 px-4 text-base hover:bg-green-700 focus:ring-green-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;

exports[`LinkButton theme should have the correct warning styles 1`] = `
<a
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-yellow-600 text-white py-2 px-4 text-base hover:bg-yellow-700 focus:ring-yellow-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  href="/somewhere"
>
  Hello
</a>
`;
